OpenRA.Mods.Common/Activities/DonateExperience.cs

```diff
@@ -0,0 +1,53 @@
+#region Copyright & License Information
+/*
+ * Copyright 2007-2017 The OpenRA Developers (see AUTHORS)
+ * This file is part of OpenRA, which is free software. It is made
+ * available to you under the terms of the GNU General Public License
+ * as published by the Free Software Foundation, either version 3 of
+ * the License, or (at your option) any later version. For more
+ * information, see COPYING.
+ */
+#endregion
+
+using OpenRA.Activities;
+using OpenRA.Mods.Common.Effects;
+using OpenRA.Mods.Common.Traits;
+using OpenRA.Traits;
+
+namespace OpenRA.Mods.Common.Activities
+{
+	class DonateExperience : Enter
+	{
+		readonly Actor target;
+		readonly int level;
+		readonly int playerExperience;
+
+		public DonateExperience(Actor self, Actor target, int level, int playerExperience)
+			: base(self, target, EnterBehaviour.Dispose)
+		{
+			this.target = target;
+			this.level = level;
+			this.playerExperience = playerExperience;
+		}
+
+		protected override void OnInside(Actor self)
+		{
+			if (target.IsDead)
+				return;
+
+			target.TraitOrDefault<GainsExperience>().GiveLevels(level);
```

Comment:
You should replace the `TraitOrDefault` by just `Trait` here: `target.Trait<GainsExperience>().GiveLevels(level);`
(`TraitOrDefault` would only return `null` if it didn't find anything, and that'd crash then because of the `.GiveLevesls`.)